### ISSUE TITLE
Support Flutter App

### ISSUE THREAD

**Issue by medz** [medz/dart-orm#328]
Prisma Dart Client was originally built for the Dart server-side, and we look forward to introducing it in Flutter.

# Progress

- [x] iOS #364
- [x] Android #370
- [x] #379
- [ ] Linux - Planned
- [ ] Windows - Planned
- [ ] Web - Planned

**Comment by listepo** [comment]
Can not wait:)

**Comment by medz** [comment]
@listepo 
Thank you for looking forward to it, but I don't know how long it will take. Prisma officials have revealed that they are experimenting with the C ABI engine, but have not announced the time or plan.

It is not feasible to implement the C ABI engine as a third party because it has already failed once. After the build is completed, the upstream repo is Prisma's internal engine. They have many destructive commits every day, making it difficult for the downstream to maintain it continuously (especially if it is only me. case)

In order to achieve the unification of Dart Server-side and Flutter, WASM is the best choice. But the real problem is that Dart does not have the ability to load WASM natively. Flutter only experimentally allows Flutter code to be compiled into wasm.

Many helplessness and many problems are like mountains before our eyes, making it difficult to overcome.

**Comment by medz** [comment]
https://github.com/prisma/prisma-engines/pull/4774

The latest news, the C ABI engine of Prisma Engines. Once it is Merge, I will immediately implement the support of Flutter.

**Comment by medz** [comment]
news https://github.com/prisma/prisma-engines/pull/4793

**Comment by listepo** [comment]
is there any success?

**Comment by medz** [comment]
@listepo 
<img width="707" alt="image" src="https://github.com/medz/prisma-dart/assets/5564821/b8acf159-5f37-49be-ab16-1963f4af7e8a">
It has not been merged yet. Once it is merged to produce the C ABI engine, I will follow up immediately.

**Comment by medz** [comment]
@listepo I noticed that it has been merged, thank you for bringing this matter to my attention again.

I am following up and looking for any information about the C ABI engine and official examples of its use in RN.

At present, it seems to only support SQLite, but it is sufficient, and most Flutter developers do not want to link to remote databases. SQLite+Prisma can help a large number of developers complete local data structure storage and have intuitive CRUD APIs

**Comment by listepo** [comment]
@medz I'll be happy to help with testing

**Comment by kidusdev** [comment]
Hy @medz i am using prisma-dart for my flutter desktop application. I didn't struggle or have error until now. How can i help?

![20240417_104930.jpg](https://github.com/medz/prisma-dart/assets/156700000/787368a6-a031-45f0-a340-7083e9204e4e)

![20240417_105050.jpg](https://github.com/medz/prisma-dart/assets/156700000/afd44844-cbe8-4822-8a02-3ae6dc57c064)


All the above data table, numbers and authentication is implemented using prisma-dart.

**Comment by medz** [comment]
Bad news, I still need to implement a C ABI engine myself. But Prisma has provided a great template

**Comment by listepo** [comment]
@medz How is it going? help is needed?

**Comment by MulverineX** [comment]
> still need to implement a C ABI engine myself

why is that? In what way is theirs unsuited for this library? I'm also curious what your timeline is looking like, and am interested in helping. After getting https://github.com/openwebf/mercury fully working for my usecase, this library is what I'm waiting on to continue with https://github.com/RefractureMedia/refracture-music

**Comment by MulverineX** [comment]
Another potential route this project could take is use of https://github.com/fzyzcjy/flutter_rust_bridge , which appears to be reaching a reliable, efficient, and cross-platform state.

**Comment by medz** [comment]
@MulverineX I hope that C Engine can also be applied on the server.

**Comment by MulverineX** [comment]
What do you mean? `flutter_rust_bridge` supports server-side Dart applications as well

**Comment by medz** [comment]
@MulverineX `flutter_rust_bridge` is useful, but my plan is to write a general C Engine. flutter_rust_bridge will keep my code locked to a certain language, internally I'm using Zig. So I may also release a Zig version of Prisma in the future.

Prisma is officially experimenting with a static version of Android/iOS (only supports sqlite), because the binary engine in the current version has many problems, such as inconsistent messages and the inability to detect the exit of the main process and terminate the binary.

I'm writing it manually in Rust and, if nothing unexpected happens, will probably release it in two months. (I'm experimenting in a private repo, not in this repo to avoid polluting commits)

**Comment by listepo** [comment]
Is it possible to run the code on only `MaOS` with `flutter`?

**Comment by medz** [comment]
@listepo APIs will not change, but customizing the Engine may be a disaster. But usually the Engine is provided for use by the Generator.

**Comment by listepo** [comment]
Is it possible to add support for the desktop version? like pure dart version for OS? load correct `prisma-query-engine` 
now error 
```
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Exception: No query engine binary found (prisma-query-engine) in [/Users/listepo/Library/Containers/com.example.flutterApplication/Data, /Users/listepo/Library/Containers/com.example.flutterApplication/Data/prisma, /Users/listepo/Library/Containers/com.example.flutterApplication/Data/.dart_tool]
```
put correct file into `/Users/listepo/Library/Containers/com.example.flutterApplication/Data` or etc

**Comment by listepo** [comment]
because for the operating system there is no difference, the code is launched only through dart or flutter

**Comment by listepo** [comment]
@medz can you give access to a private repository? I would like to help

**Comment by MulverineX** [comment]
So would I

**Comment by medz** [comment]
News:
https://github.com/prisma/prisma/issues/24295
https://github.com/dart-lang/native/issues/934#issuecomment-1894151501

My implementation: I originally built the C ABI in Rust, but upstream `prisma/prisma-engines` had too many breaking changes for me to safely maintain (I had to keep an eye on upstream and fix it). Later, I used zig combined with prisma C ABI to create a unified engine for Android/iOS. I gave up because not all developers had zig installed on their computers. Finally, I created the flutter native plug-in directly. The problem I encounter now is that the prisma c abi engine is not a dylib, but a static library. Dart ffi does not support static libraries.

Now I have two options:
1. Use c/cpp link prisma c abi to re-create the APIs
2. Use the corresponding platform language to implement it through flutter channel on different platforms.

Pros and cons:

1: The client can be implemented uniformly to reduce maintenance difficulty, but the performance will be greatly reduced.
2: Through flutter channel, call c abi in the native language, no need to worry about performance issues. Channel message uses string for data interaction.

Worrying: Prisma C ABI engine only supports sqlite3, which is definitely enough as a local database for flutter. But the server-side still has to face the previous problems.

**Comment by medz** [comment]
Now, I am working on the `flutter` branch, and I am going to sleep now, because I have been confused about the static library problem for several days, and finally found the problem, dart FFI does not support static libraries.

I will try more options tomorrow.

**Comment by medz** [comment]
On iOS, another headache is that `vendored_frameworks` in the podspec is handled by flutter tools instead of cocoapods. This caused my configuration to not work. Maybe I'll try writing a shell to solve it. I hope so. I have encountered too many problems with flutter and dart ffi recently. Most of the issues found in the search were from a few years ago and there is no solution.

**Comment by MulverineX** [comment]
In my opinion, we should just go with Option 1 until https://github.com/prisma/prisma/issues/24295 is resolved

**Comment by MulverineX** [comment]
@listepo Would you like to hop into a voice call at some point to dig into this?

**Comment by medz** [comment]
@MulverineX I experimented by adding a static library for vendored_frameworks to podcasts. Unable to find typing in Dart FFI, but can correctly link static libraries and find functions in the c file.

That is to say, a bridge needs to be built. Even if the bridge is beyond doubt, the only purpose is to compile it into a dynamic library and then call the static library of Prisma C-Abi.

**Comment by medz** [comment]
News: Flutter support has achieved initial success. I use C to call the Prisma C-ABI engine static library functions, and then ffigen hides the static library functions, resulting in a dynamic library (its only purpose is to discover the Prisma QueryEngine static library and create APIs to call it)

The source code can now be viewed at https://github.com/medz/prisma-dart/tree/flutter/packages/prisma_flutter. Dart writing for APIs has not yet been completed. Currently, only one function for creation and initialization has been written (` create`) can be seen at https://github.com/medz/prisma-dart/blob/flutter/packages/prisma_flutter/lib/src/query_engine.dart#L29.

An example of creating an engine can be viewed at https://github.com/medz/prisma-dart/blob/flutter/packages/prisma_flutter/example/lib/main.dart#L45

Currently, I am only developing on iOS support and solving the problem of Dart FFI not being able to discover third-party static libraries. (Native language code C/C++/Swift/Java can all find third-party static libraries, so you can use your own c files to perform function training)

The C-bridge code can be found at https://github.com/medz/prisma-dart/blob/flutter/packages/prisma_flutter/src/prisma_flutter.c.

This is delightful news.

**Comment by MulverineX** [comment]
![Screenshot_20240528-101401.png](https://github.com/medz/prisma-dart/assets/12068027/bdb36325-8079-4a03-9b30-1518e96dd0f5)

These still need to happen, is this work that can be done in this repo? If that's the case, I'd like to help. This issue shouldn't be closed until all 5 native platforms are supported (WASM can be a separate issue)

**Comment by medz** [comment]
News: I released `orm` v4.2.0 and `orm_flutter` v0.0.1 versions, and the integration documentation can be found at https://prisma.pub/getting-started/flutter.html, and I placed the demo program at https://github.com/medz/prisma-dart/tree/main/examples/flutter_with_orm
﻿
Next time I will start building support for Android.

**Comment by medz** [comment]
@MulverineX 

> These still need to happen, is this work that can be done in this repo?

Yes, Linux/macOS/Windows can re-package the binary engine, initialize and download the binary engine during the running of the app.

> This issue shouldn't be closed until all 5 native platforms are supported (WASM can be a separate issue).

Yes, the issue will not be closed until support for all native platforms except web is complete (sometimes the merge PR may be closed automatically and I will open it manually)

> Of course, once iOS & Android support are in, an update can be published to allow those who are developing mobile-only apps to use it.

I won't release `v1.x` until all native platforms support it, and I will be releasing on `v0.x` for a long time to come.

**Comment by MulverineX** [comment]
![Screenshot_20240528-101401.png](https://github.com/medz/prisma-dart/assets/12068027/bdb36325-8079-4a03-9b30-1518e96dd0f5)

These still need to happen, is this work that can be done in this repo? If that's the case, I'd like to help. This issue shouldn't be closed until all 5 native platforms are supported (WASM can be a separate issue)

**Comment by MulverineX** [comment]
Ideally, A version of the engine could be bundled with the app, to avoid requiring an online install. If that's how it works initially though, that's okay. Could you point me in the right direction, where I should work on getting initial support for desktop platforms in?

**Comment by medz** [comment]
@MulverineX The desktop architecture is too complex, and each architecture corresponds to an engine. Therefore, during the application startup phase, it is most reasonable to determine whether an engine needs to be downloaded and enter the download waiting page. Otherwise, your desktop app might be very large.

For the mobile side, whether including library files in the app or downloading library files is completed by the developer before building, and the packaging will only include necessary files.

If Prisma provided a dynamic library, this problem would not exist. (But judging from the Prisma roadmap, it is unlikely to be possible for a long time)

**Comment by MulverineX** [comment]
Flutter requires per arch build regardless, most of the time. What added complexity is there here? 

https://github.com/KRTirtho/spotube/releases/tag/v3.6.0

**Comment by medz** [comment]
@MulverineX Maybe you're right, it seems complex architecture only happens on Linux. macOS/Windows can include binaries directly. It seems that a special flutter.assets converter needs to be written to solve the schema file problem on Linux. Maybe I’m overcomplicating my thoughts😊

**Comment by medz** [comment]
<img width="1409" alt="image" src="https://github.com/medz/prisma-dart/assets/5564821/c0df01ff-05c8-49e9-80f8-d6894cdec7ef">

News: Support flutter Android done.

**Comment by listepo** [comment]
Nice

**Comment by listepo** [comment]
@medz maybe it makes sense to rename from `orm` to `prisma_client` and `orm_flutter` to `prisma_client_flutter`?

**Comment by MulverineX** [comment]
@listepo I think the goal here is to improve this library to the point where it can be the defacto ORM for Flutter. It wasn't already a taken name, and Prisma is generally the best option out there. Obviously, the engine still needs some query crafting work to improve performance down the road. Still, since Prisma has corporate sponsors/investment I'm confident the team will continue to iterate. Also, since this library directly interacts with the query engine, `client` isn't exactly appropriate. `prisma` being in the main description is good enough for searchability.

**Comment by medz** [comment]
@listepo I originally intended to do this, but the names are all taken (and not actually used). I tried to contact the author and the dart team through the pub.dev guide, but no emails were returned.

**Comment by medz** [comment]
@listepo Another fact is that the prisma marketing team does not want third-party clients to use short prisma-related names. So you will see that the documentation website of prisma orm for dart was updated to `Prisma Client Dart` not long ago.

**Comment by listepo** [comment]
@medz how to create and use migrations?

**Comment by medz** [comment]
@listepo Creating migrations is the same as Prisma, using the `prisma migrate dev` command.

To apply migrations, call the `applyMigrations` method on the engine

Documentation is here https://prisma.pub/getting-started/flutter.html#migrations

**Comment by listepo** [comment]
@medz thanks

**Comment by listepo** [comment]
I got error:
Got dependencies in `/Users/listepo/GitHub/monorepo/client_app`.
Error: Generator "dart run orm" failed:

Could not find package `orm` or file `orm`

**Comment by MulverineX** [comment]
Really looking forward to desktop platforms here, once they're available I can really begin to iterate on my app, as its targeting desktop first.

**Comment by listepo** [comment]
`internally I'm using Zig. So I may also release a Zig version of Prisma in the future` its very intresting

**Comment by kidusdev** [comment]
> `internally I'm using Zig. So I may also release a Zig version of Prisma in the future` its very intresting

that would be amazing if Zig took off in the future.

**Comment by listepo** [comment]
Any news?

**Comment by medz** [comment]
> Any news?

@listepo Sorry, no new news. Maybe https://github.com/prisma/prisma-engines/pull/4913 is one of the news.

My original plan was to use the binary engine to build macOS/Windows/Linux support (of course, it is not a high priority) but I did not try my best to implement it for the following reasons:

1. Prisma engine has been working towards the WASM build goal

2. The binary engine has been phased out since last year (most of the third-party language implementations are still in use)

3. There is no clear direction for Prisma to move in. I can only work as hard as possible on the engine provided by Prisma, because any implementation of my own can be easily destroyed by the above. (This is not the scope of Prisma's official obligations)

4. My personal energy is really limited. To put it bluntly, China's current unemployment rate is very high and everyone is working hard to make a living, otherwise they will not be able to eat tomorrow. (I am no exception)

Combining the above, it is difficult for Prisma Dart to make new breakthroughs. Completing similar implementations for most of the Prisma official engine is the best I can do at this point.

**Comment by MulverineX** [comment]
1. This could be very powerful, but it does come with a small performance hit
2. Im not sure what this means, but dont those third party implementations all rely on the binary engine in some way?
3. Completely understandable, keeping up with constant internal changes would be a massive pain and is infeasible for one person to maintain.
4. My most sincere condolences, its very difficult to make fulfilling & productive career/hobby choices in such a tight job market. I can sympathize a little, as I've had a tough time finding a software company to take a chance on me. I'm lucky to be living with my parents still though. It seems that most of the world is headed into a recession. I wish I could donate more and/or that some startups would take notice of this project and fund its development 😮‍💨. Good luck out there, hoping that you find some stability and rest soon.

**Comment by medz** [comment]
@MulverineX 

Thank you for your understanding and support.

> This could be very powerful, but it does come with a small performance hit

This means that it becomes more difficult to use the Prisma engine in environments that do not support loading WASM natively. Currently, other third-party languages have strong official or community support for loading WASM modules, but Dart does not. The Dart official team is currently committed to using Dart to build WASM modules and has no plans to load WASM modules at the moment.

> Im not sure what this means, but dont those third party implementations all rely on the binary engine in some way?

Yes, currently third-party languages are using the Prisma binary engine. But almost all Prisma official clients are no longer used.

As far as I know, the Wasm version of the last Scheme Engine in Prisma's official engine is about to be built. Perhaps after the last remaining binary engine implemented WASM migration, several versions of bug fixes were made. Prisma official will no longer maintain or even completely stop building binary engines (this is just my guess, because continuing to maintain binary engines after all switches to WASM is meaningless for Prisma official and will increase costs)

**Comment by MulverineX** [comment]
> strong official or community support for loading WASM modules, but Dart does not.

luckily this is actually not the case!
 https://github.com/juancastillo0/wasm_run

**Comment by listepo** [comment]
yes native code better

**Comment by medz** [comment]
@listepo Unfortunately, I deleted it after Prisma released the CABI engine. Because its essence is to use Rust FFI to write C ABI, and then call it from Zig to build a unified engine.
﻿
I am sorry that I deleted it after trying this technology (because I gave up the huge upstream dependency work)

Currently, the only thing left of Prisma Dart is the client based on the official Prisma engine.
﻿
Again, I am sorry that I no longer have the energy to try new technologies. My current personal plan for Prisma Dart (not disclosed) is as follows:

1. Stability maintenance (I will fix it once someone submits a bug)

2. Remove `code_buinder` from the code base (very low priority, I have always planned to remove `code_builder`/`build_runner`/`json_serializable`, and it took me a long time to successfully remove the `build_*` ecosystem package)

3. Migrate `fromJson` and `toJson` of <model>/<input>/<output> to macros. (Once `macros` enters the official version of Dart SDK, the experimental flag will not be required)
﻿
And my company has already given up using the Zig language. Forgive me for being too busy to do anything else. I can only keep up with the normal bug fixes for the open source project. (Because my company has once again reached the point where I am the only one left, I must find a way to save it and ensure my livelihood)

**Comment by medz** [comment]
News：
> 🎉Now [@Prisma](https://x.com/prisma) Dart v5.2.1 is released with [@FlutterDev](https://x.com/FlutterDev) integration v0.4, using a new package architecture:
> ![image](https://github.com/user-attachments/assets/cde902aa-f380-4110-8e11-b84113944a9c)
>
> 🌲We can easily implement more platform integrations in the future.

**Comment by medz** [comment]
@kidusdev Hey, the way you integrate Flutter with Prisma in Windows is exactly what I will do next to support macOS, WIndows, and Linux. Interested in contributing it?

There is already a ready-made implementation of the binary engine in the `orm` package, and the integration of the remaining three platforms is just to make it easier for Flutter Desktop to get the engine (just depend on it)

In the `orm_flutter` package, it just determines the platform and forwards the engine instance.

This request is because I need to complete more complex web integration, which usually uses WASM and uses lower-level APIs.

**Comment by medz** [comment]
Closing as superseded by the new Dart-native ORM merged in #485. This issue belongs to the retired Prisma-based client or its earlier roadmap. The new implementation uses Dart model declarations, independent drivers and single-engine Dart migrations, with no Prisma engine dependency.

The old implementation is no longer being developed. This closure does not claim that every requested feature or platform has been implemented. Please open a new issue against the new implementation with a current reproduction or focused proposal; see the repository README and capability documentation for the supported scope.

Thank you for the report and discussion.
